### PR TITLE
Add --passphrase env:VAR

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

**Comment by gperciva** [comment]
We've had a bunch of requests for non-stdin password input recenty (https://github.com/Tarsnap/scrypt/issues/183, https://github.com/Tarsnap/scrypt/issues/201, https://github.com/Tarsnap/scrypt/issues/215).  I'm not convinced that this is a good idea, but I'm also not convinced that it **isn't** a good idea.

Securely reading a passphrase from a file might be a bit messy, but reading from an environment variable looks like a simple `getenv()` call.  And if the passphrase is in the environment, presumably security isn't a huge deal.

Draft PR for comments.  The real PR would add it to `usage()`, the `README.md`, etc.

File: main.c [hunk 1]
```diff
@@ -118,6 +119,17 @@ main(int argc, char *argv[])
 				exit(1);
 			}
 			break;
+		GETOPT_OPT("--passphrase-environ"):
+			if ((passwd_environ = getenv("SCRYPT_PASSPHRASE"))
```
Comment (gperciva) [inline main.c]: On second thought, we might as well make this `GETOPT_OPTARG` and read from `optarg` here.

**Comment by gperciva** [comment]
I also have a branch with a proof-of-concept `--password-file`, in case you'd like to see that for comparison.

File: main.c [hunk 2]
```diff
@@ -188,7 +194,18 @@ main(int argc, char *argv[])
...
 
 	/* Get the password. */
-	if (0) {
+	if (passwd_env_arg != NULL) {
+		/* We're not allowed to modify the output of getenv(). */
+		if ((passwd_env = getenv(passwd_env_arg)) == NULL) {
+			warn0("Failed to read from ${%s}", passwd_env_arg);
+			goto err1;
+		}
+
```
Comment (gperciva) [inline main.c]: Should we do any sanity testing here?  for example, if somebody accidentally points to an empty environment variable like this?

```
FOO="" ./scrypt enc --passphrase-env FOO secret.txt secret.enc
```

I'm tempted to insist that the password be at least 1 character when reading from `--passphrase-env`, just to make sure that they didn't accidentally give the wrong `$VAR` or that their system didn't default to an empty string for a missing environment variable.  But that would go against the unix / C principle of "allow people to shoot their own feet".

Comment (cperciva) [reply main.c]: I'm inclined to allow zero-character environment variables.  Any system which silently converts an undefined variable to an empty one is going to run into all sorts of issues; for example, GNU's POSIXLY_CORRECT behaviours kick in even if the variable is defined to be empty.

Also, there's probably going to be scripts which use scrypt to do things and users who say "I don't actually *want* this encrypted".

**Comment by cperciva** [comment]
I'm thinking that it might be worth (somewhat) following OpenSSL's example here:

```
       env:var
           Obtain the password from the environment variable var. Since the
           environment of other processes is visible on certain platforms
           (e.g. ps under certain Unix OSes) this option should be used with
           caution.

       file:pathname
           The first line of pathname is the password. If the same pathname
           argument is supplied to -passin and -passout arguments then the
           first line will be used for the input password and the next line
           for the output password. pathname need not refer to a regular file:
           it could for example refer to a device or named pipe.
```

In the case of `openssl enc` the option name is `-pass` but I would make this a proper long option of course.

OpenSSL also supports `-pass pass:password` (which is horrible and we should not do), and `-pass fd:number`, which we might want to do at some point (but I'd wait until someone needs it).

**Comment by gperciva** [comment]
If you really want to follow their example, should we also include
```
pass:password
```
?

**Comment by gperciva** [comment]
sorry, still waking up.  Ignore that last comment.

**Comment by gperciva** [comment]
Closing this approach, as we'll include this as `--passphrase env:VAR` once #221 is in.

File: main.c [hunk 3]
```diff
@@ -116,6 +121,7 @@ main(int argc, char *argv[])
 	struct scryptdec_file_cookie * C = NULL;
 	enum passphrase_entry passphrase_entry = PASSPHRASE_UNSET;
 	const char * passphrase_arg;
+	char * passwd_env;
```
Comment (cperciva) [inline main.c:124]: Even though `getenv` returns a `char *`, I would make this `const char *` in order to enforce the "can't be modified" constraint.

**Comment by cperciva** [comment]
Looks good.

**Comment by gperciva** [comment]
Ready for merge.

**Pull request merged**